0.50 regression: Default option c_std not applied when cross-compiling

When providing `default_options: ['c_std=c99']` to the `project`, it is not being applied on `meson build --cross-file <cross-file>` invocation on a simple test project using flexible array members:

`ninja -v`

> [1/2] cc -Itest-executable@exe -I. -I../../work -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -g -pedantic -MD -MQ 'test-executable@exe/main.c.o' -MF 'test-executable@exe/main.c.o.d' -o 'test-executable@exe/main.c.o' -c ../../work/main.c
../../work/main.c:5:9: warning: ISO C90 does not support flexible array members [-Wpedantic]
     int array[];
         ^
[2/2] cc  -o test-executable 'test-executable@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed

However, after doing any kind of (even unrelated) configuration such as `meson configure -Db_asneeded=true`, the configuration is being fixed and applies as expected:

`ninja -v`

> [1/2] cc -Itest-executable@exe -I. -I../../work -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch **-std=c99** -g -pedantic -MD -MQ 'test-executable@exe/main.c.o' -MF 'test-executable@exe/main.c.o.d' -o 'test-executable@exe/main.c.o' -c ../../work/main.c
[2/2] cc  -o test-executable 'test-executable@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed

Note: I haven't tested whether this also applies to other default options.

---

#### main.c

```c
#include <stdio.h>

struct a {
    int foo;
    int array[];
};

int main(int argc, char* argv[]) {
    printf("Hey!\n");
    return 0;
}
```

#### meson.build

```meson
project('test-project', 'c', version: '0.0.1', default_options: ['c_std=c99'])
add_project_arguments('-pedantic', language: 'c')
executable('test-executable', 'main.c')
```

#### cross-file

```
[host_machine]
system = 'linux'
cpu_family = 'x86_64'
cpu = 'x86_64'
endian = 'little'

[properties]
root = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64'
needs_exe_wrapper = true

[paths]
prefix = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot'

[binaries]
c = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android28-clang'
cpp = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android28-clang++'
ar = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android-ar'
as = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android-as'
ld = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld'
strip = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android-strip'
pkgconfig = '/usr/bin/pkg-config'
```

--- a/mesonbuild/coredata.py
+++ b/mesonbuild/coredata.py
@@ -592,7 +592,13 @@ class CoreData:
                         k in env.cmd_line_options):
                     # TODO think about cross and command-line interface.
                     o.set_value(env.cmd_line_options[k])
-                self.compiler_options[for_machine].setdefault(k, o)
+                if k in ['c_std', 'cpp_std']:
+                    # These CLI options should affect all platforms
+                    for m in iter(MachineChoice):
+                        self.compiler_options[m].setdefault(k, o)
+                else:
+                    # By default, CLI options just affect the native platform
+                    self.compiler_options[for_machine].setdefault(k, o)
 
             # Unlike compiler and linker flags, preprocessor flags are not in
             # compiler_options because they are not visible to user.

--- a/run_unittests.py
+++ b/run_unittests.py
@@ -5043,6 +5043,14 @@ c = ['{0}']
         # TODO should someday be explicit about build platform only here
         self.init(testdir)
 
+    def test_std_remains(self):
+        # C_std defined in project options must be in effect also when native compiling.
+        testdir = os.path.join(self.unit_test_dir, '50 std remains')
+        self.init(testdir)
+        compdb = self.get_compdb()
+        self.assertRegex(compdb[0]['command'], '-std=c99')
+        self.build()
+
 def should_run_cross_arm_tests():
     return shutil.which('arm-linux-gnueabihf-gcc') and not platform.machine().lower().startswith('arm')
 
@@ -5093,6 +5101,14 @@ class LinuxCrossArmTests(BasePlatformTests):
                 self.assertEqual(i['value'], 'lib')
                 return
         self.assertTrue(False, 'Option libdir not in introspect data.')
+
+    def test_std_remains(self):
+        # C_std defined in project options must be in effect also when cross compiling.
+        testdir = os.path.join(self.unit_test_dir, '50 std remains')
+        self.init(testdir)
+        compdb = self.get_compdb()
+        self.assertRegex(compdb[0]['command'], '-std=c99')
+        self.build()
 
 
 def should_run_cross_mingw_tests():

--- a/test cases/unit/50 std remains/meson.build
+++ b/test cases/unit/50 std remains/meson.build
@@ -1,0 +1,2 @@
+project('std_remains', 'c', default_options: ['c_std=c99'])
+executable('prog', 'prog.c')

--- a/test cases/unit/50 std remains/prog.c
+++ b/test cases/unit/50 std remains/prog.c
@@ -1,0 +1,1 @@
+int main(int argc, char **argv) { return 0; }

